platform-sdk/docs/core/tipset-algorithm.md

```diff
@@ -0,0 +1,260 @@
+There is a special threshold that causes the snapshot to advance. The threshold for each node is slightly different,
+unless nodes have the same consensus weight. To calculate the threshold, find the minimum weight required to have >2/3
+of the total weight, and then subtract the node's weight.
+
+```
+Example of computing snapshot advancement threshold
+
+Suppose an address book has the following weights: A = 5, B = 9, C = 11, D = 2. Compute the snapshot advancement
+threshold for node A.
+
+Total consensus weight is 27. The amount of stake required to have >2/3 of the total weight is 19. Subtract A's weight,
+and A's snapshot advancement threshold is 14.
+```
+
+When a new event has a snapshot improvement score that is greater than or equal to the snapshot advancement threshold,
+then the snapshot is updated to the tipset of that new event.
+
+# Limiting Event Creation by Snapshot Improvement Score
+
+Nodes are not permitted to create a new event unless one of the following two conditions are true:
+
+1) this is the first event a node has ever created (i.e. this is genesis), or
+2) the snapshot improvement score exceeds the snapshot improvement score of the previous event, or
+3) the snapshot improvement score is greater than zero and this is the first event created since
+   we last updated the snapshot
+
+By following this rule, this ensures that each node stops creating new events in a finite amount of time if that
+node no longer is in communicating with a quorum of its peers.
+
+```
+Example
+
+Address book weights: A = 5, B = 9, C = 11, D = 2
+Consider from A's perspective. A's snapshot advancement threshold is 14.
+
+Initially, the snapshot is [-1, -1, -1, -1].
+
+A creates a genesis event with tipset [0, -1, -1, -1]. The snapshot improvement score is 0, which is ok since
+this is a genesis event. (It will never again be ok to have an advancement score of 0.)
+
+Next, A creates an event with self parent [0, -1, -1, -1] and other parent [-1, 2, -1, -1] with generation 2.
+The tipset of the new event is [3, 2, -1, -1]. The partial weighted snapshot advancement score is 9.
+
+Next, A creates an event with self parent [3, 2, -1, -1] and other parent [3, 2, 5, 3] with generation 3. The tipset
+of the new event is [4, 2, 5, 3]. The partial weighted snapshot advancement score is 22, which exceeds the threshold
+of 14. Therefore, the snapshot is updated to [4, 2, 5, 3].
+
+When it next comes time for A to create an event, it will start comparing against this new snapshot [4, 2, 5, 3].
+```
+
+# Optimizing Tipset Score
+
+When it comes time to create a new event, if there are multiple other parents to choose from, select the other parent
+that results in the greatest tipset score improvement.
+
+# Bully Scores
+
+Informally, bullying is when the network decides (for whatever reason) not to use an event creator's events as other
+parents very often. The node being bullied (the nerd) will then find that it is difficult or impossible for its events
```

Comment:
I do not know if bullying and nerd should be used for the description. At least in Germany the words are only used for toxic behavior of people that is seen as something really negative. Maybe we can find a better wording?

Reply:
Instead of bully/nerd, how about dominant/shy?

Reply:
I agree that bully/nerd is not clear languaging.    We're wanting to prevent a node from being neglected or ignored, whether by intent or by chance.   Perhaps this is another formal metric that every node has, something to prevent starvation of its events from participating in consensus.